python/lsst/ts/observatory/control/base_group.py

```diff
@@ -0,0 +1,541 @@
+# This file is part of ts_observatory_control.
+#
+# Developed for the LSST Telescope and Site Systems.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+
+import enum
+import types
+import asyncio
+import logging
+import traceback
+
+from lsst.ts import salobj
+
+__all__ = ["BaseUsages", "BaseGroup"]
+
+
+class BaseUsages(enum.IntEnum):
+    """Enumaration defining base class usages. When subclassing `BaseGroup`
+    it is recommended to define new enumerations.
+
+    Notes
+    -----
+
+    Usages definition:
+
+    All: Enable all possible operations defined in the class. This is
+    different than adding all topics from all components. It will only
+    add those needed to run the methods defined in the class.
+
+    StateTransition: Enable summary state transition.
+
+    MonitorState: Enable monitoring of summary state.
+
+    MonitorHeartBeat: Enable monitoring of heartbeat.
+    """
+
+    All = 0
+    StateTransition = 1
+    MonitorState = 1 << 1
+    MonitorHeartBeat = 1 << 2
+
+
+class BaseGroup:
+    """High-level abstraction of a collection of `salobj.Remote`.
+
+    A `salobj.Remote` provides a communication abstraction for Commandable
+    SAL Componentes (CSCs). CSCs represents entities in the LSST observatory
+    system that performs actions such as controlling/interfacing with hardware,
+    logical operations and/and coordination of other CSCs. In general CSCs
+    operate in logical groups to achieve certain kinds of operation; like
+    slewing the telescope and taking images.
+
+    Parameters
+    ----------
+    components : `list` [`str`]
+        A list of strings with the names of the components that are part
```

Comment:
"components" -> "SAL components".

Do you allow the SAL index to be part of the name? It would be a good idea to do so, and very easy to implement. (The only hard part is documenting that here).